File: Java/ReverseWordsInStr.java
public class Solution {
    public String reverseWords(String s) {
        if (s.length() < 1)
            return s;
        StringBuilder sb = new StringBuilder();
        int slow = 0, fast = s.length()-1;
        while (slow <= fast && s.charAt(slow) == ' ')
            slow++;
        while (fast >= slow && s.charAt(fast) == ' ')
            fast--;
        while (slow <= fast) {
            int cursor = fast;
            while (cursor >= slow && s.charAt(cursor) != ' ')
                cursor--;
            sb.append(s.substring(cursor+1, fast+1)+" ");
            while (cursor >= slow && s.charAt(cursor) == ' ')
                cursor--;
            fast = cursor;
        }
        String result = sb.toString();
        if (result.length()>1)
            result = result.substring(0, result.length()-1);
        return result;
        
    }
}

File: Python/ReverseWordsInStr.py
class Solution(object):
    # pointer style
    def reverseWords(self, s):
        """
        :type s: str
        :rtype: str
        """
        if s is None or len(s) < 1:
            return s
        left, right = 0, len(s)-1
        while left < len(s) and s[left] == ' ':
            left += 1
        while right >= 0 and s[right] == ' ':
            right -= 1
        result = ''
        while right >= left:
            cursor = right
            while cursor >= left and s[cursor] != ' ':
                cursor -= 1
            result += s[cursor+1:right+1] + ' '
            while cursor >= left and s[cursor] == ' ':
                cursor -= 1
            right = cursor
        return result[:-1]

    # Python list style    
    def reverseWords(self, s):
        """
        :type s: str
        :rtype: str
        """
        if s is None or len(s) < 1:
            return s
        s = s.strip().split(' ')
        result = ''
        for i in s[::-1]:
            if i != '':
                result += i + ' '
        return result[:-1]        
